DuckDuckGo/NewTabPage/Features/RecentActivityProvider.swift

```diff
@@ -0,0 +1,283 @@
+        )
+    }
+
+    let activityPublisher: AnyPublisher<[NewTabPageDataModel.DomainActivity], Never>
+
+    let historyCoordinator: HistoryCoordinating
+    let urlFavoriteStatusProvider: URLFavoriteStatusProviding
+    let urlFireproofStatusProvider: URLFireproofStatusProviding
+    let duckPlayerHistoryEntryTitleProvider: DuckPlayerHistoryEntryTitleProviding
+    let trackerEntityPrevalenceComparator: TrackerEntityPrevalenceComparing
+
+    init(
+        historyCoordinator: HistoryCoordinating,
+        urlFavoriteStatusProvider: URLFavoriteStatusProviding,
+        urlFireproofStatusProvider: URLFireproofStatusProviding,
+        duckPlayerHistoryEntryTitleProvider: DuckPlayerHistoryEntryTitleProviding,
+        trackerEntityPrevalenceComparator: TrackerEntityPrevalenceComparing
+    ) {
+        self.historyCoordinator = historyCoordinator
+        self.urlFavoriteStatusProvider = urlFavoriteStatusProvider
+        self.urlFireproofStatusProvider = urlFireproofStatusProvider
+        self.duckPlayerHistoryEntryTitleProvider = duckPlayerHistoryEntryTitleProvider
+        self.trackerEntityPrevalenceComparator = trackerEntityPrevalenceComparator
+
+        activityPublisher = historyCoordinator.historyDictionaryPublisher
+            .receive(on: DispatchQueue.main)
+            .compactMap { [weak historyCoordinator] _ -> BrowsingHistory? in
+                historyCoordinator?.history
+            }
+            .compactMap { [weak urlFavoriteStatusProvider] history -> [NewTabPageDataModel.DomainActivity]? in
+                guard let urlFavoriteStatusProvider else {
+                    return nil
+                }
+                return Self.calculateRecentActivity(
+                    with: history,
+                    urlFavoriteStatusProvider: urlFavoriteStatusProvider,
+                    urlFireproofStatusProvider: urlFireproofStatusProvider,
+                    duckPlayerHistoryItemTitleProvider: duckPlayerHistoryEntryTitleProvider,
+                    trackerEntityPrevalenceComparator: trackerEntityPrevalenceComparator
+                )
+            }
+            .eraseToAnyPublisher()
+    }
+
+    private static func calculateRecentActivity(
+        with browsingHistory: BrowsingHistory,
+        urlFavoriteStatusProvider: URLFavoriteStatusProviding,
+        urlFireproofStatusProvider: URLFireproofStatusProviding,
+        duckPlayerHistoryItemTitleProvider: DuckPlayerHistoryEntryTitleProviding,
+        trackerEntityPrevalenceComparator: TrackerEntityPrevalenceComparing
+    ) -> [NewTabPageDataModel.DomainActivity] {
+        guard !browsingHistory.isEmpty else {
+            return []
+        }
+
+        var activityItems = [DomainActivityRef]()
+        var activityItemsByDomain = [String: DomainActivityRef]()
+
+        let oneWeekAgo = Date.weekAgo
+
+        browsingHistory.filter { !$0.failedToLoad && $0.lastVisit > oneWeekAgo }
```

Comment:
This code is adapted from `HomePage.Models.RecentlyVisitedModel.refreshWithHistory`, and made testable.